tools/integration_tests/concurrent_operations/concurrent_listing_test.go

```diff
@@ -33,20 +44,486 @@ func (s *concurrentListingTest) Setup(t *testing.T) {
+func createDirectoryStructureForTestCase(t *testing.T, testCaseDir string) {
+	operations.CreateDirectory(path.Join(testDirPath, testCaseDir), t)
+
+	// Create explicitDir structure
+	explicitDir := path.Join(testDirPath, testCaseDir, "explicitDir")
+	operations.CreateDirectory(explicitDir, t)
+	operations.CreateFileOfSize(5, path.Join(explicitDir, "file1.txt"), t)
+	operations.CreateFileOfSize(10, path.Join(explicitDir, "file2.txt"), t)
+}
+
 ////////////////////////////////////////////////////////////////////////
 // Test scenarios
 ////////////////////////////////////////////////////////////////////////
 
-// TODO: Add test scenarios here.
-func (s *concurrentListingTest) TestMock1(t *testing.T) {
-	t.Log("running mock test 1")
+// Test_OpenDirAndLookUp helps in detecting the deadlock when
+// OpenDir() and LookUpInode() request for same directory comes in parallel.
+func (s *concurrentListingTest) Test_OpenDirAndLookUp(t *testing.T) {
+	t.Parallel() // Mark the test parallelizable.
+	testCaseDir := "Test_OpenDirAndLookUp"
+	createDirectoryStructureForTestCase(t, testCaseDir)
+	targetDir := path.Join(testDirPath, testCaseDir, "explicitDir")
+	var wg sync.WaitGroup
+	wg.Add(2)
+	// Fails if the operation takes more than timeout.
+	timeout := 5 * time.Second
+	iterationsPerGoroutine := 100
+
+	// Goroutine 1: Repeatedly calls OpenDir.
+	go func() {
+		defer wg.Done()
+		for i := 0; i < iterationsPerGoroutine; i++ {
+			f, err := os.Open(targetDir)
+			assert.Nil(t, err)
+
+			err = f.Close()
+			assert.Nil(t, err)
+		}
+	}()
+
+	// Goroutine 1: Repeatedly calls Stat.
+	go func() {
+		defer wg.Done()
+		for i := 0; i < iterationsPerGoroutine; i++ {
+			_, err := os.Stat(targetDir)
+			assert.Nil(t, err)
+		}
+	}()
+
+	// Wait for goroutines or timeout.
+	done := make(chan bool, 1)
+	go func() {
+		wg.Wait()
+		done <- true
+	}()
+	select {
+	case <-done:
+		// Operation completed successfully before timeout.
+	case <-time.After(timeout):
+		assert.FailNow(t, "Possible deadlock")
+	}
+}
+
+// Test_Parallel_ReadDirAndLookUp tests for potential deadlocks or race conditions when
+// ReadDir() is called concurrently with LookUp of same dir.
+func (s *concurrentListingTest) Test_Parallel_ReadDirAndLookUp(t *testing.T) {
+	t.Parallel() // Mark the test parallelizable.
+	testCaseDir := "Test_Parallel_ReadDirAndLookUp"
+	createDirectoryStructureForTestCase(t, testCaseDir)
+	targetDir := path.Join(testDirPath, testCaseDir, "explicitDir")
+	var wg sync.WaitGroup
+	wg.Add(2)
+	timeout := 200 * time.Second
+	iterationsPerGoroutine := 40
+
+	// Goroutine 1: Repeatedly calls Readdir
+	go func() {
+		defer wg.Done()
+		for i := 0; i < iterationsPerGoroutine; i++ {
+			f, err := os.Open(targetDir)
+			assert.Nil(t, err)
+
+			_, err = f.Readdirnames(0)
+			assert.Nil(t, err)
+
+			err = f.Close()
+			assert.Nil(t, err)
+		}
+	}()
+
+	// Goroutine 2: Repeatedly stats
+	go func() {
+		defer wg.Done()
+		for i := 0; i < iterationsPerGoroutine; i++ {
```

Comment:
40 stat calls will take significantly less time than ReadDir. Do you want to do unequal number of iterations in this case? Can we check this for other test cases too where operations in go-routines are different.

Reply:
Yes, thanks!